docs/modules/otaupgrade.md

```diff
@@ -0,0 +1,176 @@
+
+Wipes the spare application partition and prepares to receive the new
+application firmware.
+
+#### Syntax
+`otaupgrade.commence()`
+
+#### Parameters
+None.
+
+#### Returns
+`nil`
+
+A Lua error may be raised if the OTA upgrade cannot be commenced for some
+reason (such as due to incorrect partition setup).
+
+
+# otaupgrade.write(data)
+
+Write a chunk of application firmware data to the correct partition and
+location. Data must be streamed sequentially, the IDF does not support
+out-of-order data as would be the case from e.g. bittorrent.
+
+#### Syntax
+`otaupgrade.write(data)`
+
+#### Parameters
+- `data` a string of binary data
+
+#### Returns
+`nil`
+
+A Lua error may be raised if the data can not be written, e.g. due to the
+data not being a valid OTA image (the IDF performs some checks in this
+regard).
+
+
+# otaupgrade.complete(reboot)
+
+Finalises the upgrade, and optionally reboots into the new application
+firmware right away.
+
+#### Syntax
+`otaupgrade.complete(reboot)`
+
+#### Parameters
+- `reboot` 1 to reboot into the new firmware immediately, nil to keep running
+
+#### Returns
+`nil`
+
+A Lua error may be raised if the image does not pass validation, or no data
+has been written to the image at all.
+
+
+# otaupgrade.accept()
+
+When the installed boot loader is built with rollback support, a new
+application image is by default only booted once. During this "test run"
+it can perform whatever checks is appropriate (like testing whether it
+can still reach the update server), and if satisfied can mark itself
+as valid. Without being marked valid, upon the next reboot the system
+would "roll back" to the previous version instead.
+
+#### Syntax
+`otaupgrade.accept()`
+
+#### Parameters
+None.
+
+#### Returns
+`nil`
+
+
+# otaupgrade.rollback()
+
+If the installed boot loader is built with rollback support, a new
+firmware may decide that it is not performing as expected, and request
+an explicit rollback to the previous version. If the call to this
+function succeeds, the system will reboot without returning from the
+call.
+
+Note that it is also possible to roll back to a previous firmware
+version even after the new version has called `otaupgrade.accept()`.
+
+#### Syntax
+`otaupgrade.rollback()`
+
+#### Parameters
+None.
+
+#### Returns
+Never. Either the system is rebooted, or a Lua error is raised (e.g. due
+to there being no other firmware to roll back to).
+
+
+# otaupgrade.info()
+
+The boot info and application state and version info can be queried with
+this function. Typically it will be used to check the version of the
+running application.
+
+#### Parameters
+None.
+
+#### Returns
+A list of three values:
+- the name of the partition of the running application
+- the name of the partition currently marked for boot next (typically the
+  same as the running application, but after `otaupgrade.complete()` it
+  may point to a new application partition.
+- a table with entries for each OTA partition, containing:
```

Comment:
"a table whose keys are the names of OTA partitions and corresponding values are tables containing:" or such?

Reply:
Good idea, can do!